nixos/filesystems: switch to exfatprogs if applicable kernel

vs49688 opened the pull request:
###### Motivation for this change

See https://github.com/NixOS/nixpkgs/issues/104430#issuecomment-869167296

Fixes #104430.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

dotlambda commented on nixos/doc/manual/from_md/release-notes/rl-2111.section.xml:
```diff
@@ -331,6 +331,15 @@
+      <listitem>
+        <para>
+          If <literal>exfat</literal> is included in
+          <literal>boot.supportedFilesystems</literal> and using a
+          kernel less than 5.7, the <literal>exfatprogs</literal>
```
```suggestion
          <literal>boot.supportedFilesystems</literal> and when using
          kernel 5.7 or later, the <literal>exfatprogs</literal>
```

Infinisil commented on nixos/modules/tasks/filesystems/exfat.nix:
```diff
@@ -1,11 +1,13 @@
-
-    system.fsPackages = [ pkgs.exfat ];
-
+    system.fsPackages = (if lib.versionOlder "5.7" then [
```
`lib.versionOlder` needs two arguments.

vs49688 replied:
This'll teach me for programming tired... Will fix shortly

Infinisil commented on nixos/modules/tasks/filesystems/exfat.nix:
```diff
@@ -4,8 +4,10 @@ with lib;
-
-    system.fsPackages = [ pkgs.exfat ];
-
+    system.fsPackages = with config.boot.kernelPackages; (if kernelOlder "5.7" then [
```
I'd just use
```suggestion
    system.fsPackages = if config.boot.kernelPackages.kernelOlder "5.7" then [
```

The `with ;` expression has some problems, see https://nix.dev/anti-patterns/language#with-attrset-expression, and in this case it's not really necessary.

vs49688 committed:
nixos/filesystems/exfat: use "exfatprogs" instead of "exfat" if kernel > 5.7

5.7+ comes with a native exfat implementation, exfatprogs should be used instead.

The exfat package puts a "mount.exfat" binary in the path, which causes
mount to prefer the FUSE version to the non-fuse one. There's no way to
disable the binary, so switch to exfatprogs.

vs49688 committed:
nixos/doc: add note about switching to exfatprogs

vs49688 commented:
Ping on this?

SuperSandro2000 commented:
Please rebase the changelog.

vs49688 commented:
> Please rebase the changelog.

Done!

vs49688 commented:
@SuperSandro2000 Rebased and fixed conflicts again. Also, there's a leftover merge marker in the docs, so I've added another patch to remove it.

vs49688 commented:
Ping. Again.
